New package: StateSpaceAnalysis v0.1.0

JuliaRegistrator opened the pull request:
- Registering package: StateSpaceAnalysis
- Repository: https://github.com/harrisonritz/StateSpaceAnalysis.jl
- Created by: @harrisonritz
- Version: v0.1.0
- Commit: 136c0b99e98025884132039a9ff2ff8fa4a682d7
- Reviewed by: @harrisonritz
- Reference: https://github.com/harrisonritz/StateSpaceAnalysis.jl/issues/2#issuecomment-2518728177
- Description: Alpha version of StateSpaceAnalysis.jl
<!-- bf0c69308befbd3ccf2cc956ac8a4671113b158fd87a6d4e741b0173dab051a48f7dc43e593ffff4357bc647a3a835982385488b93f990c5e4a1a4e278a2ffe5bdf714fd207fcf87078a706de3065ec94d83d934ef78cbb8151d45cc470cf48742e30a87593d62b25fa1d3d0054bd013b1bae1576ccea7f4d1b80548301350890a21d71b28f446dad3bf34678e62bc497e050c52f5eb49a23b874fc7a4aad197037475eed0b3af06bd3315b49d9e93e17028f988d1cf3082c9e6df1a5b09c317 -->

gdalle commented:
Hi @harrisonritz, congrats on the new package! Can I ask why the existing options in Julia were not sufficient for your purposes? I know of several projects with similar scopes

harrisonritz commented:
> Hi @harrisonritz, congrats on the new package! Can I ask why the existing options in Julia were not sufficient for your purposes? I know of several projects with similar scopes

Hi @gdalle! The goal of StateSpaceAnalysis.jl is to create a method thats optimized for neuroimaging analysis. While there are some good general-purpose packages (e.g., ControlSystemsAnalysis.jl or StateSpace.jl), my package is tailored for a couple of key needs for the neuroimaging community:

* **event-related structure of Neuroimaging data**. EEG/MEG often has batched sequences (e.g., states x timesteps x trials). We are custom-built for that case by (A) including spline bases for inputs and (B) re-using the filtered/smoothed covariance across batches to massive reduce compute time.

* **high-dimensional systems**. We are designed around scaling through efficient memory allocation, robust covariance handling (via `PDMats.jl`), and regularization. This has allowed this packaged to work well for high-dimensional state space models (e.g., factor dim > observation dim).

* **lack of domain knowledge**. Because using SSMs for task-based neuroimaging is relatively new, it is difficult to provide good initializations for state space models. My packages includes modified SSID functions from `ControlSystemsAnalysis.jl` (appropriately credited and consistent with their license -- incredibly grateful!). By use SSID, we get really good initializations, which are critical for high-dimensional SSMs.

To further improve the ability for this package to benefit the neuroimaging community, my roadmap includes building out support for:
1. interfacing with a popular neuroimaging package via `PyMNE.jl`, which in the short term will allow easier importing of standard neuroimaging formats
2. building out methods for source reconstruction (localizing activity in the brain using SSMs).

This is my first Julia package, so I'm eager for suggestions & PRs. It accompanies [a recent paper](https://doi.org/10.1101/2024.09.29.615736) that uses this method in EEG. I have several research groups interested in utilizing this method for their own reserach. Since Julia is new to many in the neuroimaging community, I'm hoping that a tailored package will encourage uptake & Julia adoption.

Right now, I have to fix the compat to resubmit, and I'm going to tidy up the readme a bit.

JuliaRegistrator committed:
New package: StateSpaceAnalysis v0.1.0

UUID: 8767b432-2e83-436f-aa62-e8e2db78ab85
Repo: https://github.com/harrisonritz/StateSpaceAnalysis.jl.git
Tree: fa61b713bbd682ded5d818f955db2999eb0fa07c

Registrator tree SHA: 17aec322677d9b81cdd6b9b9236b09a3f1374c6a

harrisonritz commented:
Hi @gdalle, I have revised the readme to better reflect the targeted use-case of this package.

briochemc commented:
[noblock] My uneducated 2 cents: If this is tailored for neuroimaging, wouldn't you want "NeuroImaging" in the name?

harrisonritz commented:
[noblock] hi @briochemc -- thats a good question!

I considered NeuralStateSpace at some point (NeSS; smash bros mascot), but I didn't want to cast it too narrow either, as psychology/neuroscience is a big tent. For example, the features that make it useful for neuroscience would also be helpful in other domains, such as physiological modeling (e.g., eye-tracking, skin response, pupil dilation, facial reactions) or behavioral modeling (e.g., body movement, emotional changes, etc). In my field, some people can also be touchy about calling things 'neural' when you're getting indirect measurements from EEG/MEG/fMRI (🙄).

gdalle commented:
[noblock]

Sounds good to me! But next time be sure to register your Julia package (and thus get this kind of community feedback) _before_ you mention its name in a paper 😉

harrisonritz commented:
[noblock] ha! Yes, fair enough. Assumed I'd just host a repo before I started reading into the package development process.
